### PR TITLE
[#922] Calculate correct position (end) for issues

### PR DESCRIPTION
**Pull request by cdietrich**
[#922] Calculate correct position (end) for issues
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend [hunk 1]
```diff
@@ -337,11 +338,32 @@ import static org.eclipse.xtext.diagnostics.Severity.*
...
 			val length = (issue.length ?: 0)
+			
+			val endLineNumber = if (issue instanceof IssueExtension) {
+				(issue.endLineNumber ?: 1) - 1
+			} else {
+				lineNumber
+			}
+			
+			val endColumn = if (issue instanceof IssueExtension) {
+				if (issue.endColumn === null || issue.endColumn == -1) {
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend]: `(issue.endColumn ?: -1) == -1`

File: org.eclipse.xtext/src/org/eclipse/xtext/diagnostics/AbstractDiagnostic.java [hunk 2]
```diff
@@ -50,6 +50,15 @@ public int getColumn() {
...
 		return 0;
 	}
+	
+	public int getEndColumn() {
+		INode node = getNode();
+		if (node != null) {
+			LineAndColumn lineAndColumn = NodeModelUtils.getLineAndColumn(node, getOffset()+getLength());
+			return lineAndColumn.getColumn();
+		}
+		return 0;
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/diagnostics/AbstractDiagnostic.java]: -1 or length ?

File: org.eclipse.xtext/src/org/eclipse/xtext/validation/DiagnosticConverterImpl.java [hunk 3]
```diff
@@ -49,6 +49,17 @@ public IssueLocation() {}
...
 		public Integer offset;
 		public Integer length;
+		
+		/**
+		 * 1-based line number.
+		 * @since 2.19
+		 */
+		public Integer endLineNumber;
+		/**
+		 * 1-based column.
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/validation/DiagnosticConverterImpl.java]: Incl or excl?

File: org.eclipse.xtext/src/org/eclipse/xtext/validation/Issue.java [hunk 4]
```diff
@@ -18,6 +18,25 @@
...
+	 * @author Christian Dietrich - Initial contribution and API
+	 * @since 2.19
+	 */
+	interface IssueExtension {
+		/**
+		 * Returns the column in the end line of the issue. It's not the virtual column but literally
+		 * the character offset in the column, e.g. tab ('\t') counts as one character.
+		 * The first char in a line has column number 1, the number is one-based.
+		 * 
+		 * If no column information is available, returns -1.
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/validation/Issue.java]: My gut feeling is, that `default getEndColumn() { return -1 }` would fit nicely into Issue itself.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend (hunk 1 above)
Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend]: :+1:

File: org.eclipse.xtext/src/org/eclipse/xtext/diagnostics/AbstractDiagnostic.java (hunk 2 above)
Comment (cdietrich) [reply org.eclipse.xtext/src/org/eclipse/xtext/diagnostics/AbstractDiagnostic.java]: hmmm this is analogous to getColumn

File: org.eclipse.xtext/src/org/eclipse/xtext/validation/DiagnosticConverterImpl.java (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext/src/org/eclipse/xtext/validation/DiagnosticConverterImpl.java]: no idea. this is why i dislike this

File: org.eclipse.xtext/src/org/eclipse/xtext/validation/Issue.java (hunk 4 above)
Comment (cdietrich) [reply org.eclipse.xtext/src/org/eclipse/xtext/validation/Issue.java]: and what for the endlinenumber?

Comment (cdietrich) [reply org.eclipse.xtext/src/org/eclipse/xtext/validation/Issue.java]: a getLineNumber?

File: org.eclipse.xtext/src/org/eclipse/xtext/validation/DiagnosticConverterImpl.java (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext/src/org/eclipse/xtext/validation/DiagnosticConverterImpl.java]: == last included char

**Comment by cdietrich** [comment]
we decided to go with the first variant for now

**Pull request closed without merging**